### ISSUE TITLE
Devel::CheckLib failing on valid config.

### ISSUE THREAD

**Issue by toddr** [perl5-dbi/DBD-mysql#109]
Previously working and bisected to the introduction of Devel::CheckLib in 9eda4cc9870ef97a6fa61ddce6e26f318c8e7dfb

I'm using the mariadb-native-client modules for linking from mysql.so.

perl Makefile.PL INSTALLDIRS=vendor --mysql_config=/usr/local/cpanel/3rdparty/bin/mariadb_config 

When I run, I get:

```
/usr/local/cpanel/3rdparty/perl/524/bin/perl Makefile.PL INSTALLDIRS=vendor --mysql_config=/usr/local/cpanel/3rdparty/bin/mariadb_config --testuser=root --testsocket=/var/lib/mysql/mysql.sock
/usr/local/cpanel/3rdparty/bin/mariadb_config: unrecognized option '--embedded'
I will use the following settings for compiling and testing:

  cflags        (mysql_config ) = -I/usr/local/cpanel/3rdparty/include/mariadb -Wunused -Wno-uninitialized
  embedded      (guessed      ) = 
  ldflags       (guessed      ) = 
  libs          (mysql_config ) = -L/usr/local/cpanel/3rdparty/lib64/mariadb -lmariadb -lpthread -lssl -ldl -lm
  mysql_config  (User's choice) = /usr/local/cpanel/3rdparty/bin/mariadb_config
  nocatchstderr (default      ) = 0
  nofoundrows   (default      ) = 0
  nossl         (default      ) = 0
  testdb        (default      ) = test
  testhost      (default      ) = 
  testpassword  (default      ) = 
  testport      (default      ) = 
  testsocket    (User's choice) = /var/lib/mysql/mysql.sock
  testuser      (User's choice) = root

To change these settings, see 'perl Makefile.PL --help' and
'perldoc DBD::mysql::INSTALL'.

Checking if libs are available for compiling...
/home/rpmbuild/rpm-build/BUILD/DBD-mysql-4.042/assertlibLotOK4qE: error while loading shared libraries: libmariadb.so.1: cannot open shared object file: No such file or directory
wrong result: 'mariadb'

**Comment by toddr** [comment]
Patching out Devel::CheckLib works for me. I suspect whatever is processing --mysql_config isn't making it in to the check? I'll investigate.

**Comment by pali** [comment]
> I'm using the mariadb-native-client modules for linking from mysql.so.

Which version? Some versions of MariaDB are known to be broken, see reported bug: https://jira.mariadb.org/browse/MDEV-12056

>   libs          (mysql_config ) = -L/usr/local/cpanel/3rdparty/lib64/mariadb -lmariadb -lpthread -lssl -ldl -lm
>  /home/rpmbuild/rpm-build/BUILD/DBD-mysql-4.042/assertlibLotOK4qE: error while loading shared libraries: libmariadb.so.1: cannot open shared object file: No such file or directory

Which means that library `libmariadb.so.1` expanded from param `-lmariadb` is not available in system lib directories (/lib /usr/lib/ ...) and also not in `/usr/local/cpanel/3rdparty/lib64/mariadb` specified by your mysql_config script `/usr/local/cpanel/3rdparty/bin/mariadb_config`.

So recheck where you have your libmariadb.so stored, if it is correct symlink and if script supplied correct path...

**Comment by toddr** [comment]
```
$>ls -ld /usr/local/cpanel/3rdparty/lib64/mariadb/libmariadb.so.1
-rwxr-xr-x 1 root root 1269985 Nov 20  2015 /usr/local/cpanel/3rdparty/lib64/mariadb/libmariadb.so.1

**Comment by pali** [comment]
Please add `debug => 1` option into `assert_lib()` call in Makefile.PL. It should provide verbose debug output.

**Comment by toddr** [comment]
It doesn't seem to be adding the lib path right?

```
perl Makefile.PL INSTALLDIRS=vendor --mysql_config=/usr/local/cpanel/3rdparty/bin/mariadb_config --testuser=root --testsocket=/var/lib/mysql/mysql.sock
/usr/local/cpanel/3rdparty/bin/mariadb_config: unrecognized option '--embedded'
I will use the following settings for compiling and testing:

  cflags        (mysql_config ) = -I/usr/local/cpanel/3rdparty/include/mariadb -Wunused -Wno-uninitialized
  embedded      (guessed      ) = 
  ldflags       (guessed      ) = 
  libs          (mysql_config ) = -L/usr/local/cpanel/3rdparty/lib64/mariadb -lmariadb -lpthread -lssl -ldl -lm
  mysql_config  (User's choice) = /usr/local/cpanel/3rdparty/bin/mariadb_config
  nocatchstderr (default      ) = 0
  nofoundrows   (default      ) = 0
  nossl         (default      ) = 0
  testdb        (default      ) = test
  testhost      (default      ) = 
  testpassword  (default      ) = 
  testport      (default      ) = 
  testsocket    (User's choice) = /var/lib/mysql/mysql.sock
  testuser      (User's choice) = root

To change these settings, see 'perl Makefile.PL --help' and
'perldoc DBD::mysql::INSTALL'.

Checking if libs are available for compiling...
# Compiler seems to be /usr/bin/gcc
# /usr/bin/gcc -DPERL_DISABLE_PMC -fPIC -DPIC -I/usr/local/cpanel/3rdparty/perl/524/include -L/usr/local/cpanel/3rdparty/perl/524/lib64 -I/usr/local/cpanel/3rdparty/include -L/usr/local/cpanel/3rdparty/lib64 -m64 -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -Wl,-rpath -Wl,/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector-strong -L/usr/local/lib assertlibwuDdalOF.c -o assertlib1EVZDSxX -L/usr/local/cpanel/3rdparty/lib64/mariadb -lmariadb
/home/rpmbuild/rpm-build/BUILD/DBD-mysql-4.042/assertlib1EVZDSxX: error while loading shared libraries: libmariadb.so.1: cannot open shared object file: No such file or directory
# /usr/bin/gcc -DPERL_DISABLE_PMC -fPIC -DPIC -I/usr/local/cpanel/3rdparty/perl/524/include -L/usr/local/cpanel/3rdparty/perl/524/lib64 -I/usr/local/cpanel/3rdparty/include -L/usr/local/cpanel/3rdparty/lib64 -m64 -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -Wl,-rpath -Wl,/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector-strong -L/usr/local/lib assertlibwuDdalOF.c -o assertlibwTXP7XuF -L/usr/local/cpanel/3rdparty/lib64/mariadb -lpthread
# /usr/bin/gcc -DPERL_DISABLE_PMC -fPIC -DPIC -I/usr/local/cpanel/3rdparty/perl/524/include -L/usr/local/cpanel/3rdparty/perl/524/lib64 -I/usr/local/cpanel/3rdparty/include -L/usr/local/cpanel/3rdparty/lib64 -m64 -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -Wl,-rpath -Wl,/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector-strong -L/usr/local/lib assertlibwuDdalOF.c -o assertlibJNiLZYZS -L/usr/local/cpanel/3rdparty/lib64/mariadb -lssl
# /usr/bin/gcc -DPERL_DISABLE_PMC -fPIC -DPIC -I/usr/local/cpanel/3rdparty/perl/524/include -L/usr/local/cpanel/3rdparty/perl/524/lib64 -I/usr/local/cpanel/3rdparty/include -L/usr/local/cpanel/3rdparty/lib64 -m64 -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -Wl,-rpath -Wl,/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector-strong -L/usr/local/lib assertlibwuDdalOF.c -o assertlibDSKURBCW -L/usr/local/cpanel/3rdparty/lib64/mariadb -ldl
# /usr/bin/gcc -DPERL_DISABLE_PMC -fPIC -DPIC -I/usr/local/cpanel/3rdparty/perl/524/include -L/usr/local/cpanel/3rdparty/perl/524/lib64 -I/usr/local/cpanel/3rdparty/include -L/usr/local/cpanel/3rdparty/lib64 -m64 -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -Wl,-rpath -Wl,/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector-strong -L/usr/local/lib assertlibwuDdalOF.c -o assertlibiLBv3hMt -L/usr/local/cpanel/3rdparty/lib64/mariadb -lm
wrong result: 'mariadb'

**Comment by toddr** [comment]
```
$> echo "int main(int argc, char *argv[]) { return 0; }"  > foo.c

$> /usr/bin/gcc -DPERL_DISABLE_PMC -fPIC -DPIC -I/usr/local/cpanel/3rdparty/perl/524/include -L/usr/local/cpanel/3rdparty/perl/524/lib64 -I/usr/local/cpanel/3rdparty/include -L/usr/local/cpanel/3rdparty/lib64 -m64 -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -Wl,-rpath -Wl,/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector-strong -L/usr/local/lib foo.c -o foo -L/usr/local/cpanel/3rdparty/lib64/mariadb -lmariadb
$> ./foo
./foo: error while loading shared libraries: libmariadb.so.1: cannot open shared object file: No such file or directory
$> ldd foo
	linux-vdso.so.1 =>  (0x00007fff6e5fe000)
	libgdbm.so.4 => /lib64/libgdbm.so.4 (0x00007f3c91330000)
	libmariadb.so.1 => not found
	libc.so.6 => /lib64/libc.so.6 (0x00007f3c90f6e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3c91548000)

**Comment by toddr** [comment]
this works:

```
$> export LD_LIBRARY_PATH=/usr/local/cpanel/3rdparty/lib64/mariadb
$> ./foo
$> ldd foo
	linux-vdso.so.1 =>  (0x00007fffa1b31000)
	libgdbm.so.4 => /lib64/libgdbm.so.4 (0x00007f12ee242000)
	libmariadb.so.1 => /usr/local/cpanel/3rdparty/lib64/mariadb/libmariadb.so.1 (0x00007f12edfee000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f12edc2d000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f12eda11000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f12ed80c000)
	libssl.so.10 => /lib64/libssl.so.10 (0x00007f12ed5a0000)
	libcrypto.so.10 => /lib64/libcrypto.so.10 (0x00007f12ed1bb000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f12eceb8000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f12ee45a000)
	libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007f12ecc6e000)
	libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007f12ec98e000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f12ec789000)
	libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007f12ec554000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f12ec33e000)
	libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007f12ec12f000)
	libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007f12ebf2b000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f12ebd11000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f12ebae9000)
	libpcre.so.1 => /lib64/libpcre.so.1 (0x00007f12eb888000)

**Comment by toddr** [comment]
This is starting to look like a bug in Devel::CheckLib. I notice that when I get past this and run make, it sets LD_RUN_PATH, which Devel::CheckLib doesn't seem to be doing?

```
rm -f blib/arch/auto/DBD/mysql/mysql.so
LD_RUN_PATH="/usr/local/cpanel/3rdparty/lib64/mariadb:/lib64" /usr/bin/gcc  -shared -Os -L/usr/local/cpanel/3rdparty/perl/524/lib64 -L/usr/local/cpanel/3rdparty/lib64 -L/usr/lib64 -L/lib64 -L/usr/local/lib -fstack-protector-strong dbdimp.o mysql.o socket.o  -o blib/arch/auto/DBD/mysql/mysql.so 	\
   -L/usr/local/cpanel/3rdparty/lib64/mariadb -lmariadb -lpthread -lssl -ldl -lm  	\

**Comment by toddr** [comment]
Looks like @salva beat me to the analysis: https://github.com/mattn/p5-Devel-CheckLib/issues/19

**Comment by pali** [comment]
From your output it is for sure problem with rpath. And really path specified in env LD_RUN_PATH for gcc is stored in output ELF binary as RPATH.

So I think it is bug in Devel::CheckLib too.

But we could workaround it and set LD_RUN_PATH env in perl before calling assert_lib().

Do you have idea how to make or makemaker fill LD_RUN_PATH?

**Comment by toddr** [comment]
EU::MM works fine. Devel::CheckLib is just reporting failure when there is none.

**Comment by pali** [comment]
Yes EU::MM works fine, but if tou find out how EU::MM fills LD_RUN_PATH env, then we can use same algorithm also for Devel::CheckLib or setting $ENV{LD_RUN_PATH} before calling assert_lib().

**Comment by pali** [comment]
It should be fixed in Devel::CheckLib 1.08. In CHANGES is:
```
1.08    2017-03-18      Add $LD_RUN_PATH for the libs.
```
